From 843c9e13f3363cf7f4ee131b13270e55cb9a2778 Mon Sep 17 00:00:00 2001
From: fujiwarat <takao.fujiwara1@gmail.com>
Date: Tue, 2 Nov 2010 11:21:24 +0900
Subject: [PATCH] Replace s/gdk_drawable_get_size/gdk_window_get_height/ for GTK3.

---
 client/gtk2/ibusimcontext.c |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/client/gtk3/ibusimcontext.c b/client/gtk3/ibusimcontext.c
index 0bb71b5..c06faaa 100644
--- a/client/gtk3/ibusimcontext.c
+++ b/client/gtk3/ibusimcontext.c
@@ -669,10 +669,15 @@ _set_cursor_location_internal (GtkIMContext *context)
 
     area = ibusimcontext->cursor_area;
     if (area.x == -1 && area.y == -1 && area.width == 0 && area.height == 0) {
+#if GTK_CHECK_VERSION (2, 91, 0)
+        area.x = 0;
+        area.y += gdk_window_get_height (ibusimcontext->client_window);
+#else
         gint w, h;
         gdk_drawable_get_size (ibusimcontext->client_window, &w, &h);
         area.y += h;
         area.x = 0;
+#endif
     }
 
     gdk_window_get_origin (ibusimcontext->client_window, &x, &y);
-- 
1.7.2.1

