--- ibus-1.5.1/setup/keyboardshortcut.py.orig	2013-01-15 17:00:05.093191054 +0900
+++ ibus-1.5.1/setup/keyboardshortcut.py	2013-01-15 17:00:18.034160760 +0900
@@ -207,7 +207,7 @@ class KeyboardShortcutSelection(Gtk.VBox
     def __set_selected_shortcut(self, shortcut):
         model = self.__shortcut_view.get_model()
         path, column = self.__shortcut_view.get_cursor()
-        model[path[0]][0] = shortcut
+        model[path][0] = shortcut
         self.__update_add_and_apply_buttons()
 
     def __del_selected_shortcut(self):
