--- ibus/ui/gtk3/candidatepanel.vala.orig	2012-02-04 09:11:06.233979510 +0900
+++ ibus/ui/gtk3/candidatepanel.vala	2012-02-04 09:16:18.813558453 +0900
@@ -20,6 +20,7 @@
  * Boston, MA  02111-1307  USA
  */
 
+using Cairo;
 using Gtk;
 using Pango;
 
@@ -33,7 +34,7 @@ public class CandidatePanel : Gtk.HBox{
     private CandidateArea m_candidate_area;
     private HSeparator m_hseparator;
 
-    private Gdk.Rectangle m_cursor_location;
+    private Cairo.RectangleInt m_cursor_location;
 
     public signal void cursor_up();
     public signal void cursor_down();
@@ -80,7 +81,7 @@ public class CandidatePanel : Gtk.HBox{
     }
 
     public void set_cursor_location(int x, int y, int width, int height) {
-        Gdk.Rectangle location = { x, y, width, height };
+        Cairo.RectangleInt location = { x, y, width, height };
         if (m_cursor_location == location)
             return;
         m_cursor_location = location;
--- ibus/ui/gtk3/handle.vala.orig	2012-02-04 08:36:38.812523442 +0900
+++ ibus/ui/gtk3/handle.vala	2012-02-04 09:10:36.858017677 +0900
@@ -26,7 +26,7 @@ using Gtk;
 
 class Handle : Gtk.EventBox {
     private bool m_move_begined;
-    private Gdk.Rectangle m_workarea;
+    private Cairo.RectangleInt m_workarea;
     private Gdk.Point m_press_pos;
 
     public signal void move_begin();
@@ -153,7 +153,7 @@ class Handle : Gtk.EventBox {
             Gtk.StyleContext context = get_style_context();
             Gtk.Allocation allocation;
             get_allocation(out allocation);
-            Gtk.render_handle(context, cr,
+            context.render_handle(cr,
                 allocation.x, allocation.y + (allocation.height - 40) / 2, allocation.width, 40.0);
         }
         return false;
