From 888189eba783ac014e7c367004a70dad6f7b8ac0 Mon Sep 17 00:00:00 2001
From: Peng Huang <shawn.p.huang@gmail.com>
Date: Mon, 6 Sep 2010 17:17:24 +0800
Subject: [PATCH] Fix memory leak in ibus_input_context_process_key_event by Fujiwarat

---
 src/ibusinputcontext.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/src/ibusinputcontext.c b/src/ibusinputcontext.c
index 6755727..7359de0 100644
--- a/src/ibusinputcontext.c
+++ b/src/ibusinputcontext.c
@@ -775,6 +775,10 @@ _process_key_event_reply_cb (IBusPendingCall *pending,
                        call_data->keycode,
                        call_data->state | IBUS_FORWARD_MASK);
     }
+
+    if (reply_message != NULL) {
+        dbus_message_unref (reply_message);
+    }
 }
 
 static void
-- 
1.7.2.1

