From 50fa73713cae31a09d65f5ff4097458e76df3231 Mon Sep 17 00:00:00 2001
From: fujiwarat <takao.fujiwara1@gmail.com>
Date: Wed, 9 Feb 2011 11:05:09 +0900
Subject: [PATCH] Implement GJS UI for GNOME-Shell.

---
 configure.ac                  |    2 +
 ui/Makefile.am                |    1 +
 2 files changed, 4536 insertions(+), 0 deletions(-)

diff --git a/configure.ac b/configure.ac
index f1575e7..50aaaf5 100644
--- a/configure.ac
+++ b/configure.ac
@@ -392,6 +392,8 @@ ibus/_config.py
 ibus/Makefile
 ibus/interface/Makefile
 ui/Makefile
+ui/gjs/Makefile
+ui/gjs/_ibus/Makefile
 ui/gtk/Makefile
 ui/gtk/ibus-ui-gtk
 ui/gtk/gtkpanel.xml.in
diff --git a/ui/Makefile.am b/ui/Makefile.am
index a0e287e..2f1a0af 100644
--- a/ui/Makefile.am
+++ b/ui/Makefile.am
@@ -22,6 +22,7 @@
 
 if ENABLE_PYTHON
 SUBDIRS = \
+	gjs \
 	gtk \
 	$(NULL)
 endif
-- 
1.7.3.2

