diff --git a/ui/gtk3/panel.vala b/ui/gtk3/panel.vala
index 54251e7..69d0723 100644
--- a/ui/gtk3/panel.vala
+++ b/ui/gtk3/panel.vala
@@ -386,7 +386,7 @@ class Panel : IBus.PanelService {
             prev_version = var_prev_version.dup_string();
         }
 
-        if (compare_versions(prev_version, "1.5.3") < 0) {
+        if (compare_versions(prev_version, "1.5.2") < 0) {
             update_version_1_5_3();
         }
 
