From 85565b7a146e5ccbd484273b647b6195b6c08b03 Mon Sep 17 00:00:00 2001
From: fujiwarat <takao.fujiwara1@gmail.com>
Date: Thu, 2 Sep 2010 12:58:42 +0900
Subject: [PATCH] Fix segv in bus_engine_proxy_process_key_event_reply_cb

---
 bus/engineproxy.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/bus/engineproxy.c b/bus/engineproxy.c
index adda91c..64dda4f 100644
--- a/bus/engineproxy.c
+++ b/bus/engineproxy.c
@@ -564,7 +564,9 @@ bus_engine_proxy_process_key_event_reply_cb (IBusPendingCall *pending,
             /* reply timeout */
             IBusObject *connection;
             connection = (IBusObject *) ibus_proxy_get_connection ((IBusProxy *)call_data->engine);
-            ibus_object_destroy (connection);
+            if (connection) {
+                ibus_object_destroy (connection);
+            }
         }
         g_warning ("%s: %s", error->name, error->message);
         ibus_error_free (error);
-- 
1.7.2.1

