From fcfb99cfe6d83974c43bd41a1544d7a16460240f Mon Sep 17 00:00:00 2001
From: fujiwarat <takao.fujiwara1@gmail.com>
Date: Tue, 17 Jul 2012 14:35:17 +0900
Subject: [PATCH] Fix not to switch engines by non-trigger keys.

---
 ui/gtk3/switcher.vala |    6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/ui/gtk3/switcher.vala b/ui/gtk3/switcher.vala
index b543a8f..ea56c07 100644
--- a/ui/gtk3/switcher.vala
+++ b/ui/gtk3/switcher.vala
@@ -128,6 +128,7 @@ class Switcher : Gtk.Window {
                 state & KeybindingManager.MODIFIER_FILTER);
 
         update_engines(engines);
+        m_result = 0;
         m_selected_engine = index;
         m_label.set_text(m_buttons[index].longname);
         m_buttons[index].grab_focus();
@@ -343,6 +344,11 @@ class Switcher : Gtk.Window {
                     break;
                 default:
                     debug("0x%04x", pe->keyval);
+                    if (m_loop != null) {
+                        m_loop.quit();
+                        m_loop = null;
+                    }
+                    retval = false;
                     break;
             }
         } while (false);
-- 
1.7.10.4

